### PR TITLE
Refactor Go API of pkg/events

### PR DESCRIPTION
**Pull request by htdvisser**
#### Summary

This PR refactors the Go API of `pkg/events` in preparation for #2632.

#### Changes

- `events.Define` now takes variadic options (instead of visibility rights) and returns an `events.Builder` interface (instead of a function).
- Updated all places that define events to move visibility into option.
- Updated all places that call events (which are now structs, not funcs) to use a convenience function `NewWithIdentifiersAndData` which has the same signature as the old event functions.
- Changes in the Network Server to work with the new Builders instead of closures.

#### Testing

Existing unit tests should still pass.

##### Regressions

Can't think of anything that could break at runtime, unless I made some stupid mistake in my regex find/replace

#### Notes for Reviewers

This makes a lot of small changes in many places, so let's try to get this merged ASAP to avoid conflicts.

#### Checklist

- [ ] Scope: The referenced issue is addressed, there are no unrelated changes.
	- ⚠️ No, this is only preparation, but I think it's acceptable to do this separately.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
	- Apparently the events documentation wasn't updated for `visibility` so it's correct again.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

**Review by johanstokking — APPROVED** [review]
Looks good in the sample review that I did

@rvolosatovs to check the NS part still

**Review by adriansmares — APPROVED** [review]
LGTM but as mentioned above, leaving final word for NS changes to @rvolosatovs .

File: pkg/events/builder.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+func (b *builder) BindData(data interface{}) Builder {
+	return b.With(WithData(data))
+}
+
+// Builder is the interface for building events from definitions.
+type Builder interface {
+	With(opts ...Option) Builder
+	New(ctx context.Context, opts ...Option) Event
+
+	// Convenience function for legacy code. Same as New(ctx, WithIdentifiers(ids), WithData(data)).
```
Comment (rvolosatovs) [inline pkg/events/builder.go:77]: nit; A nicer way to implement these would be via functions, which would take `Builder` interface as parameter.
Also, these are *methods*.

File: pkg/events/options.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+		for _, ids := range identifiers {
+			e.innerEvent.Identifiers = append(e.innerEvent.Identifiers, ids.EntityIdentifiers())
+		}
+	})
+}
+
+// WithData returns an option that sets the data of the event.
+func WithData(data interface{}) Option {
+	return optionFunc(func(e *event) {
+		e.data = data
```
Comment (rvolosatovs) [inline pkg/events/options.go:49]: this should panic if `e.data` is not nil

File: pkg/events/options.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+				e.innerEvent.CorrelationIDs = mergeStrings(e.innerEvent.CorrelationIDs, cids)
+			}
+		}
+	})
+}
+
+// WithVisibility returns an option that sets the visibility of the event.
+func WithVisibility(rights ...ttnpb.Right) Option {
+	return optionFunc(func(e *event) {
+		e.innerEvent.Visibility = ttnpb.RightsFrom(rights...)
```
Comment (rvolosatovs) [inline pkg/events/options.go:63]: `panic` if already set

File: pkg/events/options.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+				authentication.TokenID = tokenID
+			}
+		}
+		if md.XForwardedFor != "" {
+			xff := strings.Split(md.XForwardedFor, ",")
+			authentication.RemoteIP = strings.Trim(xff[0], " ")
+		}
+		if authentication.RemoteIP != "" || authentication.TokenID != "" ||
+			authentication.TokenType != "" || authentication.Type != "" {
+			e.innerEvent.Authentication = authentication
```
Comment (rvolosatovs) [inline pkg/events/options.go:92]: `panic` if already set

**Review by kschiffer — APPROVED** [review]
Code owned files LGTM

File: pkg/events/options.go (hunk 2 above)
Comment (htdvisser) [reply pkg/events/options.go:49]: I'm just moving what was already there. The goal here is not to reimplement.

File: pkg/events/options.go (hunk 3 above)
Comment (htdvisser) [reply pkg/events/options.go:63]: I'm just moving what was already there. The goal here is not to reimplement.

File: pkg/events/options.go (hunk 4 above)
Comment (htdvisser) [reply pkg/events/options.go:92]: I'm just moving what was already there. The goal here is not to reimplement.

File: pkg/events/options.go (hunk 2 above)
Comment (rvolosatovs) [reply pkg/events/options.go:49]: That is not true - before `data` could only be specified once, now it's possible to specify it multiple times - you changed the semantics

File: pkg/events/options.go (hunk 3 above)
Comment (rvolosatovs) [reply pkg/events/options.go:63]: See above

**Pull request merged**